lib/os/queued_operation.c

```diff
@@ -0,0 +1,531 @@
+{
+	mgr->state = ST_IDLE;
+	select_next_and_unlock(mgr, key);
+}
+
+static inline int op_get_priority(const struct queued_operation *op)
+{
+	return (s8_t)(op->notify.flags >> QUEUED_OPERATION_PRIORITY_POS);
+}
+
+static inline int op_set_priority(struct queued_operation *op,
+				  int priority)
+{
+	s8_t prio = (s8_t)priority;
+	u32_t mask = (QUEUED_OPERATION_PRIORITY_MASK
+		      << QUEUED_OPERATION_PRIORITY_POS);
+
+	if (priority == QUEUED_OPERATION_PRIORITY_PREPEND) {
+		prio = INT8_MIN;
+	} else if (priority == QUEUED_OPERATION_PRIORITY_APPEND) {
+		prio = INT8_MAX;
+	} else if (prio != priority) {
+		return -EINVAL;
+	}
+
+	op->notify.flags = (op->notify.flags & ~mask)
+			   | (mask & (prio << QUEUED_OPERATION_PRIORITY_POS));
+
+	return 0;
+}
+
+static inline void finalize_and_notify(struct queued_operation_manager *mgr,
+				       struct queued_operation *op,
+				       int res)
+{
+	sys_notify_generic_callback cb
+		= sys_notify_finalize(&op->notify, res);
+
+	if (cb != NULL) {
+		mgr->vtable->callback(mgr, op, cb);
+	}
+}
+
+/* React to the completion of an onoff transition, either from a
+ * manager or directly.
+ *
+ * @param mgr the operation manager
+ *
+ * @param from either ST_STARTING or ST_STOPPING depending on
+ * transition direction
+ *
+ * @param res the transition completion value, negative for error.
+ */
+static void settle_onoff(struct queued_operation_manager *mgr,
+			 enum state from,
+			 int res)
+{
+	k_spinlock_key_t key = k_spin_lock(&mgr->lock);
+
+	__ASSERT_NO_MSG((mgr->state == from)
+			|| (state_is_error(mgr->state)
+			    && (ST_ERROR == from)));
+
+	if (res >= 0) {
+		if (from == ST_STARTING) {
+			trivial_start_and_unlock(mgr, key);
+			return;
+		}
+
+		/* Came from STOPPING or RESETTING so set to OFF.  If
+		 * the underlying onoff service is in an error state
+		 * that will be discovered when it is next used.
+		 */
+		mgr->state = ST_OFF;
+
+		k_spin_unlock(&mgr->lock, key);
+		return;
+	}
+
+	/* On transition failure mark service failed.  All unstarted
+	 * operations are unlinked and completed as a service failure.
+	 */
+	sys_slist_t ops = mgr->operations;
+
+	sys_slist_init(&mgr->operations);
+	mgr->state = ST_ERROR;
+
+	k_spin_unlock(&mgr->lock, key);
+
+	struct queued_operation *op;
+
+	SYS_SLIST_FOR_EACH_CONTAINER(&ops, op, node) {
+		finalize_and_notify(mgr, op, -ENODEV);
+	}
+}
+
+static void start_callback(struct onoff_manager *oomgr,
+			   struct onoff_client *cli,
+			   u32_t state,
+			   int res)
+{
+	struct queued_operation_manager *mgr
+		= CONTAINER_OF(cli, struct queued_operation_manager, onoff_client);
+
+	settle_onoff(mgr, ST_STARTING, res);
+}
+
+static void stop_and_unlock(struct queued_operation_manager *mgr,
+			    k_spinlock_key_t key)
+{
+	__ASSERT_NO_MSG(mgr->state == ST_IDLE);
+
+	if (mgr->onoff == NULL) {
+		mgr->state = ST_OFF;
+		k_spin_unlock(&mgr->lock, key);
+		return;
+	}
+
+	mgr->state = ST_STOPPING;
+
+	struct onoff_manager *onoff = mgr->onoff;
```

Comment:
seems redundant as `settle_onoff(mgr, ST_STOPPING, onoff_release(mgr->onoff));` would fit in a line.